frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/validation/PortsConsistentValidation.java

```diff
@@ -0,0 +1,91 @@
+// All rights reserved.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+// documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package com.microsoft.frameworklauncher.common.validation;
+
+import com.microsoft.frameworklauncher.common.model.Ports;
+import com.microsoft.frameworklauncher.common.model.ResourceDescriptor;
+import com.microsoft.frameworklauncher.common.model.ValueRange;
+import com.microsoft.frameworklauncher.common.utils.ValueRangeUtils;
+
+import javax.validation.Constraint;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+import javax.validation.Payload;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.ArrayList;
+import java.util.Map;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+
+@Target({FIELD})
+@Retention(RUNTIME)
+@Constraint(validatedBy = {PortsConsistentValidation.Validator.class})
+
+public @interface PortsConsistentValidation {
+
+  String message() default "{com.microsoft.frameworklauncher.common.validation.PortsConsistentValidation.message}";
+
+  Class<?>[] groups() default {};
+
+  Class<? extends Payload>[] payload() default {};
+
+  @Target({FIELD})
+  @Retention(RUNTIME)
+  @interface List {
+    PortsConsistentValidation[] value();
+  }
+
+  public static class Validator implements ConstraintValidator<PortsConsistentValidation, ResourceDescriptor> {
+    @Override
+    public void initialize(PortsConsistentValidation constraintAnnotation) {
+    }
+
+    @Override
+    public boolean isValid(ResourceDescriptor r, ConstraintValidatorContext context) {
+      // Not null is already handled by NotNull validator
+      if (r == null || r.getPortDefinitions() == null) {
+        return true;
+      }
+
+      // Convert port information from user input format to List<Range> format for AM scheduling.
+      java.util.List<ValueRange> portRangeList = new ArrayList<ValueRange>();
```

Comment:
>java.util.List [](start = 6, length = 14)

import

Reply:
Cannot import, there is another List in line 52

---
In reply to: [173688428](https://github.com/Microsoft/pai/pull/318#discussion_r173688428) [](ancestors = 173688428)